### DISCUSSION TITLE
getServerSideProps: Why is page not refreshing after navigating to the same URL with different query params?

### DISCUSSION THREAD

**Discussion by florianwalther-private** [vercel/next.js#38099] [Help] [answered]
One of my pages has this `getServerSideProps` function:

```
export const getServerSideProps: GetServerSideProps = async ({ resolvedUrl }) => {
try {
        const searchParams = new URLSearchParams(resolvedUrl.split('?')[1]);

        const response = await TutHubApi.searchResources(searchParams);

        return {
            props: {
                resultData: response,
            }
        }
    } catch (error) {
        console.log(error);
        return { props: { resourceLoadingErrorMessage: Constants.GENERIC_SERVER_ERROR_MESSAGE } }
    }
```

If the user used an invalid page number for the paginated results, I navigate them to the same page with the correct page number in the query params:

```
 useEffect(() => {
        if (currentPage < 1) {
            router.query.page = '1';
            router.push(router);
        } else if (resultData && (currentPage > resultData.pageCount)) {
            router.query.page = resultData.pageCount.toString();
            router.push(router);
        }
    }, [currentPage, resultData, router]);
``` 

**Problem: Even tho this successfully fetches the new data (I can see it in the logs), it doesn't refresh the page content. I have to refresh the page manually.**

It works if I add a `key` prop to the returned props from `getServerSideProps`. But why is this necessary and is it the correct way?

**Comment by ramirezsandin** [comment]
Why don't you use the `query` object to read the query params instead the `resolvedUrl ` inside getServerSideProps, that way it will be better.

**Reply by florianwalther-private** [reply]
Because I don't know how to URL encode the `ParsedUrlQuery` to send it to my server. I actually opened a thread about it but got no answers: https://github.com/vercel/next.js/discussions/38061

**Reply by ramirezsandin** [reply]
Basically you'll have to check the parameter you need inside router.query, and check if it is undefined, a string or an array of strings and do your call to the API with those params

```typescript
export const getServerSideProps = ({ query }: GetServerSidePropsContext) => {
  const queryParam = query['query-param-name']
  if (!queryParam) {
    // Is not in the URL
  } else if (Array.isArray(queryParam)) {
    // There are many
  } else {
    // Only one param
  }
  return {
    props: {},
  }
}
```
To encode the values you can then do `encodeURIComponent(value)` to any param you want to pass to the server

**Reply by ramirezsandin** [reply]
Also, you want to move the logic that is currently in your `useEffect` to `getServerSideProps`. That way, nothing will render if the page number is wrong, and it will redirect to the proper page just after the call to the API is made. You can use something like this

```typescript
export const getServerSideProps = ({ query }: GetServerSidePropsContext) => {
  const pageParam = query['page']

  let page = 1
  if (!pageParam) {
    // Is not in the URL
  } else if (Array.isArray(pageParam)) {
    // There are many
  } else if (!isNaN(parseInt(pageParam))) {
    // Only one param
    page = parseInt(pageParam)
  }

  const resultData = await TutHubApi.searchResources(searchParams)
  if (page < 1) {
    return {
      redirect: {
        destination: `/whatever-path-this-is?page=1`,
        permanent: false,
      },
    }
  }
  if (page > resultData.pageCount) {
    return {
      redirect: {
        destination: `/whatever-path-this-is?page=${resultData.pageCount}`,
        permanent: false,
      },
    }
  }

  return {
    props: {
      resultData,
    },
  }
}

```

**Reply by florianwalther-private** [reply]
Thank you, that's a great idea. I actually had the logic in `getServerSideProps` before, but without the redirect so the new URL didn't show the correct (updated) page param. 

With your approach of handling the query params, I have to treat each parameter individually, right? Seems like a lot of code compared to just sending all query params as they are.

**Reply by ramirezsandin** [reply]
Yes, you can treat them independently, or throw them all to the API request, but, what if someone write manually a wrong parameter, or write like: www.yoursite.com/search?page=3&page=other

Some validation should be made on the parameters before sending them elsewhere, if not, there's margin to get some error, they can be only 1 string, multiple string, numbers, characters, etc..

What I usually do is use the [Yup ](https://github.com/jquense/yup)to make the validations, and sometimes I share these validations among my own APIS, hooks, forms,... and any place you want to check values that can be modified by the user.

Another option is just treat them as the type of value that you're expecting and send them to the API call, and if any error occurs, just capture it and show it on the user interface.

**Reply by florianwalther-private** [reply]
@ramirezsandin If the query params are invalid, the server will respond with an error and the UI will reflect that. I would really like to just send all query params as they are to the server. With `UrlSearchParams` that was easy.

**Reply by icyJoseph** [reply]
What's so different about `URLSearchParams`? What did you do before? 

> I would really like to just send all query params as they are to the server.

This is exactly what you have, I am very confused now. We are side-tracking here with a bunch of implementation details, what's exactly the design you are after? Could you list it out?

**Reply by florianwalther-private** [reply]
@icyJoseph There seems to be no simple way to url-encode `ParsedUrlQuery` which makes it difficult to work with. I wonder why Next decided to use that type instead of `URLSearchParams`.

**Comment by promisehalle** [comment]
@[Bloxorz](https://bloxorz.io) let read it!
Thanks for sharing!

**Reply by ramirezsandin** [reply]
spam

**Comment by icyJoseph** [comment]
Hi,

I think you need to do something like:

```ts
  useEffect(() => {
    const currentQuery = router.query;
    if (currentPage < 1) {
      router.push({
        pathname: router.pathname,
        query: { ...currentQuery, page: "1" },
      });
    } else if (resultData && currentPage > resultData.pageCount) {
      router.push({
        pathname: router.pathname,
        query: { ...currentQuery, page: resultData.pageCount.toString() },
      });
    }
  }, [currentPage, resultData, router]);
```

`router.push(router)` is not producing a new object, IIRC, it's like setting state to the same object. When you add `key`, I guess you tell it that it is a different thing, similarly, when people want to trigger quick-and-dirty, re-render of a child, people often pass a key, and control the value of the key, changing it when a re-render is needed, where everything else (props) is the same. 

> See [here](https://codesandbox.io/s/k9znx9kj9r?file=/src/index.js), if you remove the error overlay (x to the top right corner), you can re-trigger the boundary, because they change the key.

Also this would trigger a waterfall of warnings, right? `Unknown key passed via urlObject into url.format: beforePopState`.

I think though, this has potential to trigger a runaway `useEffect`, right? Every run of this effect, changes `router`, which then triggers this, and unless the `if`-`else if` blocks are properly ignored in the second run, this will spin forever. Scary.

**Reply by florianwalther-private** [reply]
Your suggested solution doesn't work either, unfortunately.

**Reply by icyJoseph** [reply]
I noticed I misread the key thing. Would you mind sharing more of the component logic?

**Reply by ramirezsandin** [reply]
This should work, that depends on how you're making the pagination.

**Reply by icyJoseph** [reply]
I tried with, which is a simplified case I guess:

```tsx
import type { GetServerSidePropsContext, NextPage } from "next";
import { useRouter } from "next/router";
import { useEffect } from "react";

const Home: NextPage<{ page: number; results: { foo: string }[] }> = (
  props
) => {
  const router = useRouter();

  const { page } = router.query;

  const pageNumber = Number([page].join(""));

  const { page: propsPage, results } = props;

  useEffect(() => {
    if (pageNumber !== propsPage) {
      const { pathname, query } = router;
      router.push({ pathname, query: { ...query, page: propsPage } });
    }
  }, [pageNumber, propsPage, router]);

  if (Number.isNaN(pageNumber)) return <div>Error page number is NaN</div>;

  return (
    <div>
      {results.map(({ foo }) => (
        <p key={foo}>{foo}</p>
      ))}
    </div>
  );
};

export const getServerSideProps = (ctx: GetServerSidePropsContext) => {
  const query = ctx.query;

  const page = [query.page].join("");

  const pageNumber = Number(page);

  if (Number.isNaN(pageNumber)) return { notFound: true };

  if (pageNumber > 10) {
    return {
      props: {
        results: [],
        page: Math.floor(Math.random() * 11),
      },
    };
  }

  return {
    props: {
      results: Array.from({ length: pageNumber + 1 }, (_, i) => ({
        foo: `bar ${i}`,
      })),
      page: pageNumber,
    },
  };
};

export default Home;
```

It seems to do what you want? If I go to `/?page=100000` it bounces me back to something between 0 and 10, otherwise, it lets me through and I render a list with as many items as the page number + 1.

### Gotcha!

It might be better to do `router.replace` though, and remove the invalid entry from the navigation history, because in my case, going back, throws you forward to a random page 😅

**Reply by florianwalther-private** [reply]
@icyJoseph When I only change the query params, the page does not update with the newest values (unless I use the key prop).
I do get the latest values, but the component simply doesn't rerender.

**Reply by icyJoseph** [reply]
Then I need to see this component, the snippet above I'd say is the base case, and it does re-render as expected.

You could try to make a minimal repro, or try to change the snippet above until it triggers the non-rendering behaviour.

**Reply by florianwalther-private** [reply]
> Then I need to see this component, the snippet above I'd say is the base case, and it does re-render as expected.
> 
> You could try to make a minimal repro, or try to change the snippet above until it triggers the non-rendering behaviour.

Here is the full component (minus the imports). The `router.push` happens in the `updateQueryParam` function. Even if I use your approach by passing an object there, it still doesn't update the page without a key.

https://textuploader.com/t1b48

**Reply by icyJoseph** [reply]
I see that you are locked up by a problem I call, "transforming props to state":

```tsx
const ResourcesPage = ({ resultData, resourceLoadingErrorMessage, editResource, voteResource, bookmarkResource, showAuthRequest, showEmailVerificationRequest, countSearch }: ServerSideProps & PageProps) => {
    const router = useRouter();

    const searchQueryParam = router.query.q?.toString();

    const currentPage = parseInt(router.query.page?.toString() ?? "1");

    const [results, setResults] = useState<Resource[] | undefined>(resultData?.results);

```

This last line here, is problematic for me. 

I think if you want a short answer, you should do:

```tsx
 useEffect(() => {
      setResults(resultData?.results);
    }, [resultData]);
```

The broader issue being that `useState`'s initialisation argument is used once, and then never called, this is why sometimes we pass a function to it, because otherwise we'd waste resources, assigning memory to something that remains unused.

When the `getServerSideProps` result is transported as `json` to the application, it is used as input props, and since you are transforming props to state, you need to run an effect to synchronise your state with the new incoming props.

### Additional considerations

#### Key

Since you pass a new key on the JSON, that forces React to re-render the entire child, as a new child. 

> Keys are useful for component instance identity beyond lists as well. You can add a key to any React component at any time to indicate its identity, and changing that key will cause React to destroy the old component instance and DOM and create new ones... from [here](https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#keys-and-reconciliation)

#### Accumulating JSX with forEach

This is a personal thing, I would say that the, `forEach` here is not needed. Instead, define a React component that renders the list:

```tsx
const ResourceEntries = ({
  resources,
  onResourceVoted,
  onEditResourceClicked,
  showAuthRequest,
  showEmailVerificationRequest,
  onBookmarkIconClicked,
}: {
  resources: Resource[];
  onResourceVoted: (resourceId: string, upDown: string) => Promise<void>;
  /* Complete type with other props */
}) => {
  return resources.map((resource) => (
    <ResourceEntry
      key={resource.id}
      resource={resource}
      onEditResourceClicked={onEditResourceClicked}
      showAuthRequest={showAuthRequest}
      showEmailVerificationRequest={showEmailVerificationRequest}
      onResourceVoted={onResourceVoted}
      onBookmarkIconClicked={onBookmarkIconClicked}
    />
  ));
};
```

I skipped a bunch of the types, and then you'd use it like:

```tsx

  async function onResourceVoted(resourceId: string, upDown: string) {
    const updatedVotes = await voteResource(resourceId, upDown);

    if (updatedVotes) {
      const updatedResults = results.map((resource) =>
        resource._id === resourceId
          ? { ...resource, votes: updatedVotes }
          : resource
      );
      setResults(updatedResults);
    }
  }

             // later on the JSX:
             {/* include other props - reduced for simplicity */}
                <ResourceEntries
                  resources={results}
                  onResourceVoted={onResourceVoted}
                />
```

#### Bonus

This article describes something you could make use of.

https://www.joshwcomeau.com/nextjs/refreshing-server-side-props/

Basically, for your `onResourceVoted` function, you could:

```ts
async function onResourceVoted(resourceId: string, upDown: string) {
    const updatedVotes = await voteResource(resourceId, upDown);

    if (updatedVotes) {
      await router.replace(router.asPath);
    }
  }
```

And then you can get rid of the `useState`, and instead just work directly with `resultData.results`, and of course also get rid of the `useEffect`, that is, of course, if fetching the props from the server once again is plausible.

**Reply by florianwalther-private** [reply]
@icyJoseph Thank you! I use the state because when a user votes a resource, I want to reflect that in the UI. Is the `useEffect` the right way to go or is it just a hack and I should handle it in some other way?

Also, related to that, does the intermediate state negatively affect search engine crawlers? Or is it just as good as rendering the prop that I directly get from `getServerSideProps`?

**Reply by icyJoseph** [reply]
> Thank you! I use the state because when a user votes a resource, I want to reflect that in the UI. Is the useEffect the right way to go or is it just a hack and I should handle it in some other way?

Look at my updated answer, specifically the Bonus part. 

The `useEffect` is correct though, if props must be transformed into state, but it is often the case, that you can engineer it away, and derive what you need from the props, or create essential state (like filters) which are used to transform the props.

> Also, related to that, does the intermediate state negatively affect search engine crawlers? Or is it just as good as rendering the prop that I directly get from getServerSideProps?

It depends. Of course when it doesn't work for you, then you might assume that it won't work well for the bot/crawler. One thing to remember is that, as of now, it is probably best to not assume that bots run JS, though some do.

**Reply by icyJoseph** [reply]
Hope it works out! if it does please, don't forget to mark as answered!

**Reply by ramirezsandin** [reply]
> Here is the full component (minus the imports). The `router.push` happens in the `updateQueryParam` function. Even if I use your approach by passing an object there, it still doesn't update the page without a key.
> 
> https://textuploader.com/t1b48

🤯
I would recommend not to save the data returned from getServerSideProps in a state, and instead to get the data using some library like swr or react query, so the data is refetch every time the url string changes, and you can also trigger a mutate order to update the data (eg when the user click the upvote).

**Reply by florianwalther-private** [reply]
@icyJoseph @ramirezsandin  Thank you. It seems wasteful to refetch the whole page after a vote. I already get the updated item back from the server, I was hoping to just replace it client side (hence the separate state). 
Of course, just refetching the data would make my life easier. Do you think I'm overoptimizing by mutating the existing state?

**Reply by florianwalther-private** [reply]
And thank you very much @icyJoseph for the other code improvements!

**Reply by icyJoseph** [reply]
> Do you think I'm overoptimizing by mutating the existing state?

Nah, not at this stage, but you see the issues it can create. One has to be mindful about those. 

Some might say though, that the real optimisation is to fly the request, and update the data optimistically on the client, without awaiting for the response and instead just assume the update will work. [Optimistic UI update](https://medium.com/@_erikaybar/optimistic-ui-updates-in-react-9e139ffa2e45).

Once the update is done, you finish that async work with reconciliation, either by refetching the server props, or by using the response from the to synch the local state to what the server says is the truth. 

```ts
// this:
updateLikeCount(comment).then((likes) => setState(likes));
setState(comment.likes + 1);
// and not this:
const likes = await updateLikeCount(comment);
setState(likes + 1);
```


This is of course not written in stone, some APIs by design choice of back-end engineers, might just return 204's, or a result of the operation, but not the entire updated document. 

> Hard lined Command-Query-Separation, would require a query to get the update results, and return nothing, or just an status report, from the update command.

I think, at the end of the day, the right implementation, as long as it meets the design criteria, sits somewhere between the easiest way, and the one that makes most sense to you and your team.

**Reply by icyJoseph** [reply]
Last but not least, please don't forget to mark answered when you are done. If you ended up doing something entirely different, post it and make that the accepted answer.

**Reply by florianwalther-private** [reply]
@icyJoseph @ramirezsandin Thank you for all your help!

**Comment by icyJoseph** [comment, marked as answer]
In short the answer to OP is to:

- keep state derived from props, synched by using a useEffect

Alternatively, in this case, `getServerSideProps` props can be simply re-fetched by doing, `router.replace(router.asPath)` when the user actions change the value of the props in a remote server, as done here, https://www.joshwcomeau.com/nextjs/refreshing-server-side-props/

In this case, introducing yet another library to fetch data from the client, would also require at least another API route, and overall increases the complexity.
